### PR TITLE
semconv: Add metric generation

### PR DESCRIPTION
**Pull request by carrbs**
Closes: #4528 

This is a first attempt in generating some metric constants as a first step in completing deeper generation of semantic conventions for metrics. I borrowed a bit of the[ jinja logic from opentelemetry-python](https://github.com/open-telemetry/opentelemetry-python/blob/main/scripts/semconv/templates/semantic_metrics.j2), as well as some of the formatting functions from the jinja template used in this repos semconv directory.

After taking a look at the [RELEASE.md](https://github.com/open-telemetry/opentelemetry-go/blob/main/RELEASING.md), I'm now thinking the generated `semconv/v1.24.0/metric.go` file shouldn't be included in this PR.

### CODE REVIEW

**Comment by carrbs** [comment]
I added some more `const`s in this last commit. This would allow us to change the [usage](https://github.com/open-telemetry/opentelemetry-go-contrib/blob/main/instrumentation/google.golang.org/grpc/otelgrpc/config.go#L88C1-L90C1) mentioned in #4528 like this:

https://github.com/open-telemetry/opentelemetry-go/blob/bd60b70c32d3af917c624254540b560ca9301116/semconv/v1.24.0/metric.go#L729-L737

from `opentelemetry-go-contrib/blob/main/instrumentation/google.golang.org/grpc/otelgrpc/config.go`:

```go
// original
c.rpcDuration, err = c.meter.Float64Histogram("rpc."+role+".duration",
	metric.WithDescription("Measures the duration of inbound RPC."),
	metric.WithUnit("ms"))

// with generated consts
c.rpcDuration, err = c.meter.Float64Histogram(RPCServerDurationName,
	metric.WithDescription(RPCServerDurationDescription),
	metric.WithUnit(RPCServerDurationUnit))
```

@pellared suggested "at least the instrument name [could be generated], but maybe something that creates a whole instrument would be better? but it may be too complex". I think it's likely possible we can gather these items together, possibly in the jinja. Shouldn't be too complex. I want to get some feedback on the current implementation to make sure it's on track.

A few other notes for discussing:
- I noticed that the usage case here hardcodes `"Measures the duration of inbound RPC"` even though the `role` is a variable. (i.e., the message should be "Measures the duration of outbound RPC" when `role == 'client'`).
- I came across `metric_group` as an option for identifying a specific type of semantic convention, but I don't see any definitions (outside of some [test data in the build_tools repo](https://github.com/open-telemetry/build-tools/blob/f8b934b6f827633e251157f6b9cc834198694577/semantic-conventions/src/tests/data/yaml/metrics.yaml#L4)) utilizing it. Likely a question for semantic conventions SIG group.

```

File: CHANGELOG.md [hunk 1]
```diff
@@ -8,6 +8,10 @@ This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.htm
 
 ## [Unreleased]
 
+### Added
+
+- Adds a new template to generate metric semantic conventions, as well as adding it to the Makefile `semconv-generate` script to add the metric.go file to the v1.24.0 directory. (#4528)
```
Comment (carrbs) [inline CHANGELOG.md:13]: This will need to be changed to reflect the actual/final solution.

File: semconv/v1.24.0/metric.go [hunk 2]
```diff
@@ -0,0 +1,1084 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Code generated from semantic convention specification. DO NOT EDIT.
+
+package semconv // import "go.opentelemetry.io/otel/semconv/v1.24.0"
```
Comment (carrbs) [inline semconv/v1.24.0/metric.go:17]: I ran this with `v1.24.0`, it's probably supposed to be `v1.24.1` or `v.1.25.0`

**Comment by carrbs** [comment]
Examples of generated code and comments:
- defined `stability` and `brief` (notice the `it_reps()` output can be duplicative, adding "It represents..." to all `brief`s:
```go
	// RPCServerDuration is the metric conforming to the "rpc.server.duration"
	// semantic conventions. It represents the measures the duration of inbound
	// RPC.
	// Instrument: histogram
	// Unit: ms
	// Stability: Experimental
	RPCServerDurationName        = "rpc.server.duration"
	RPCServerDurationUnit        = "ms"
	RPCServerDurationDescription = "Measures the duration of inbound RPC."
```

- when undefined `stability`/`brief` in [semantic-conventions](https://github.com/open-telemetry/semantic-conventions/blob/main/model/metrics/system-metrics.yaml#L322):
```go
	// SystemFilesystemUsage is the metric conforming to the
	// "system.filesystem.usage" semantic conventions.
	// Instrument: updowncounter
	// Unit: By
	// Stability: Experimental
	// NOTE: The description (brief) for this metric is not defined in the semantic-conventions repository.
	SystemFilesystemUsageName = "system.filesystem.usage"
	SystemFilesystemUsageUnit = "By"
```

File: CHANGELOG.md (hunk 1 above)
Comment (carrbs) [reply CHANGELOG.md:19]: Might not need to include the version information because I don't believe the metric.go file will be included in the final version of this PR

File: semconv/v1.24.0/metric.go [hunk 3]
```diff
@@ -0,0 +1,1082 @@
...
+	DBClientConnectionsUseTimeUnit        = "ms"
+	DBClientConnectionsUseTimeDescription = "The time between borrowing a connection and returning it to the pool"
+
+	// AspnetcoreRoutingMatchAttempts is the metric conforming to the
+	// "aspnetcore.routing.match_attempts" semantic conventions. It represents the
+	// number of requests that were attempted to be matched to an endpoint.
+	// Instrument: counter
+	// Unit: {match_attempt}
+	// Stability: Experimental
+	AspnetcoreRoutingMatchAttemptsName        = "aspnetcore.routing.match_attempts"
```
[the remark this answers is not in the document]
Comment (carrbs) [reply semconv/v1.24.0/metric.go:120]: ooh, good catch! I can look into this. What are your thoughts on including the generated file in this PR @pellared? Once we get it generating correctly, should that be part of a separate (release?) PR?

File: semconv/v1.24.0/metric.go [hunk 4]
```diff
@@ -0,0 +1,1082 @@
...
+	HTTPClientResponseBodySizeName        = "http.client.response.body.size"
+	HTTPClientResponseBodySizeUnit        = "By"
+	HTTPClientResponseBodySizeDescription = "Size of HTTP client response bodies."
+
+	// JvmMemoryInit is the metric conforming to the "jvm.memory.init" semantic
+	// conventions. It represents the measure of initial memory requested.
+	// Instrument: updowncounter
+	// Unit: By
+	// Stability: Experimental
+	JvmMemoryInitName        = "jvm.memory.init"
```
[the remark this answers is not in the document]
Comment (carrbs) [reply semconv/v1.24.0/metric.go:500]: Is this something we'd want to change for the other generated conventions? For example: 

https://github.com/open-telemetry/opentelemetry-go/blob/54b6ee41746734577b513eb3e0ba4f205fc87e6a/semconv/v1.24.0/attribute_group.go#L298


https://github.com/open-telemetry/opentelemetry-go/blob/54b6ee41746734577b513eb3e0ba4f205fc87e6a/semconv/v1.24.0/attribute_group.go#L440

**Comment by carrbs** [comment]
Update:

I added some little string replacements to the jinja template for the metrics, but I believe a better approach would be to add `"ASPNETCore"` and `"JVM"` to the `capitalizations` slice in [opentelemetry-go-build-tools](https://github.com/open-telemetry/opentelemetry-go-build-tools/blob/main/semconvgen/generator.go#L285). I rebuilt the `semconvgen` and it produced the result (across all semantic conventions), for example:

```diff
+++ b/semconv/v1.24.0/attribute_group.go
@@ -284,7 +284,7 @@ func PoolName(val string) attribute.KeyValue {
 
 // ASP.NET Core attributes
 const (
-       // AspnetcoreDiagnosticsHandlerTypeKey is the attribute Key conforming to
+       // ASPNETCoreDiagnosticsHandlerTypeKey is the attribute Key conforming to
        // the "aspnetcore.diagnostics.handler.type" semantic conventions. It
        // represents the full type name of the
        // [`IExceptionHandler`](https://learn.microsoft.com/dotnet/api/microsoft.aspnetcore.diagnostics.iexceptionhandler)
```

This however, would be a breaking change, so we should discuss our options.

File: semconv/metric_template.j2 [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+{% for id in semconvs %}
+{%- if semconvs[id].GROUP_TYPE_NAME == 'metric' %}{% set metric = semconvs[id] %}
+  // {{ keydoc(metric) | wordwrap(76, break_long_words=false, break_on_hyphens=false, wrapstring="\n// ") }}
+  // Instrument: {{ metric.instrument }}
+  // Unit: {{ metric.unit }}
+  // Stability: {{ format_stability(metric.stability) }}
+{%- if not metric.brief %}
+  // NOTE: The description (brief) for this metric is not defined in the semantic-conventions repository.
+{%- endif %}
+  {{to_go_name(metric.metric_name)}}Name = "{{metric.metric_name}}"
```
Comment (carrbs) [inline semconv/metric_template.j2:48]: Should the metric name have the string "`Name`" prepended? 
i.e., which is better/easier to work with: `JVMMemoryUsedName` vs. `JVMMemoryUsed`

Comment (pellared) [reply semconv/metric_template.j2:48]: I find the current way (e.g. `JVMMemoryUsedName`) better.

File: semconv/v1.24.0/metric.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (pellared) [reply semconv/v1.24.0/metric.go]: I believe that the ones you listed above are "attributes keys" which are different than "metric names".

File: semconv/v1.24.0/metric.go (hunk 2 above)
Comment (pellared) [reply semconv/v1.24.0/metric.go:6]: `v1.24.0` is good. We can add things to the latest semconv package.

File: semconv/v1.24.0/metric.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (carrbs) [reply semconv/v1.24.0/metric.go]: If we only want to apply the capitalization to metric names (for ASPNET/JVM), it might be easier to use the jinja template changes I added yesterday (instead of adding `capitalization-path` functionality to the `semconvgen`).

**Comment by carrbs** [comment]
I see four options at this point:
1. Leave the PR as-is, limit the formatting logic of metric names to the metrics jinja template:
```jinja
{% macro to_go_name(name) -%}
    {%- if name.startswith('aspnetcore.') -%}
        ASPNETCore{{ name[11:] | replace(".", " ") | replace("_", " ") | title | replace(" ", "") }}
    {%- elif name.startswith('jvm.') -%}
        JVM{{ name[4:] | replace(".", " ") | replace("_", " ") | title | replace(" ", "") }}
    {%- else -%}
        {{ name | replace(".", " ") | replace("_", " ") | title | replace(" ", "") }}
    {%- endif -%}
{%- endmacro %}
```
2. Revert the special handling altogether and keep the formatting as `JvmFooBar` and `AspnetcoreFooBar`
3. Add `ASPNETCore` and `JVM` to the `capitalizations` slice in the `semconvgen` tool, which improves (?)  readability across all generated `semconv`.
4. Modify the `semconvgen` tool to accept a custom capitalizations file (of type ?) to append to the `capitalizations` slice:

```bash
# ... in Makefile
$(SEMCONVGEN) -i "$(OTEL_SEMCONV_REPO)/model/." --only=metric  -f metric.go \
 -t "$(SEMCONVPKG)/metric_template.j2" \
--capitalizations-path=/path/to/metric-capitalizations.txt -s "$(TAG)"
```
The third option is the only one introducing breaking changes since metric generations are being introduced in this PR. @pellared (maybe @MadVikingGod too), if you have some guidance on which approach is best, I'd appreciate your insight.

**Comment by pellared** [comment]
I think we can start with option 2 and work on 4 concurrently so that we can use the new feature when we will be generating `semconv/v1.25.0`.

**Comment by carrbs** [comment]
Ticket for adding the custom capitalizations flag to `semconvgen`: https://github.com/open-telemetry/opentelemetry-go-build-tools/issues/528

**Comment by carrbs** [comment]
[PR](https://github.com/open-telemetry/opentelemetry-go-build-tools/pull/529) (WIP) for adding the flag and processing of custom capitalizations.

**Review by pellared — APPROVED** [review]

**Comment by pellared** [comment]
@open-telemetry/go-approvers PTAL

**Review by XSAM — APPROVED** [review]

File: CHANGELOG.md [hunk 6]
```diff
@@ -139,7 +140,7 @@ See our [versioning policy](VERSIONING.md) for more information about these stab
 - Upgrade all use of `go.opentelemetry.io/otel/semconv` to use `v1.24.0`. (#4754)
 - Update transformations in `go.opentelemetry.io/otel/exporters/zipkin` to follow `v1.24.0` version of the OpenTelemetry specification. (#4754)
 - Record synchronous measurements when the passed context is canceled instead of dropping in `go.opentelemetry.io/otel/sdk/metric`.
-  If you do not want to make a measurement when the context is cancelled, you need to handle it yourself (e.g  `if ctx.Err() != nil`). (#4671)
+  If you do not want to make a measurement when the context is cancelled, you need to handle it yourself (e.g `if ctx.Err() != nil`). (#4671)
```
Comment (dmathieu) [inline CHANGELOG.md:143]: Could we extract all those changelog fixes into a dedicated PR?

Comment (carrbs) [reply CHANGELOG.md]: @dmathieu: Those changes we unintended (from a bad merge conflict resolution), I've updated the CHANGELOG.

File: CHANGELOG.md [hunk 7]
```diff
@@ -2974,3 +2975,4 @@ It contains api and sdk for trace and meter.
 [metric API]:https://pkg.go.dev/go.opentelemetry.io/otel/metric
 [metric SDK]:https://pkg.go.dev/go.opentelemetry.io/otel/sdk/metric
 [trace API]:https://pkg.go.dev/go.opentelemetry.io/otel/trace
+
```
Comment (dmathieu) [inline CHANGELOG.md:2978]:
```suggestion
```

**Review by dmathieu — APPROVED** [review]
